File: Examples/Python/Containers/MongoDB-REST/src/controllers/telemetry_controller.py
import time
import json
from flask import Blueprint, request, current_app, jsonify

telemetry_blueprint = Blueprint('telemetry', __name__)

db_name = 'test'
collection_name = 'source'

@telemetry_blueprint.route('/<int:start>', methods=['POST'])
def post(start):
    container_time = time.time()
    ssl_handshake = container_time - start
    start_ref = time.time()

    measure = time.time()
    data = request.json
    deserialization = time.time() - measure

    measure = time.time()
    client = current_app.mongodb_rest_client.get_client()
    document_id = client.create(db_name, collection_name, data)
    create_object = time.time() - measure

    measure = time.time()
    document = client.read(db_name, collection_name, document_id)
    get_object = time.time() - measure

    measure = time.time()
    json_data = json.dumps(document, default=str)
    serialization = time.time() - measure

    measure = time.time()
    for _ in range(15):
        client.update(db_name, collection_name, document_id, {'time': time.time()})
    high_iops = time.time() - measure

    measure = time.time()
    client.delete(db_name, collection_name, document_id)
    delete_object = time.time() - measure

    total = time.time() - start_ref

    telemetry = {
        'start': start,
        'container_time': container_time,
        'ssl_handshake': ssl_handshake,
        'deserialization': deserialization * 1000,
        'create_object': create_object * 1000,
        'get_object': get_object * 1000,
        'serialization': serialization * 1000,
        'high_iops': high_iops * 1000,
        'delete_object': delete_object * 1000,
        'total': total * 1000,
        'type': 'python'
    }

    return jsonify(telemetry), 200

File: Examples/Rust/Containers/MongoDB-REST/src/src/controllers/telemetry_controller.rs
use actix_web::{post, web, HttpResponse, Responder};
use serde_json::json;
use serde_json::Value;
use mongodb::{bson::{ doc }};
use serde::{Deserialize};

use chrono::{Utc, TimeZone};
use std::time::{SystemTime, UNIX_EPOCH};

use crate::services::mongodb::rest::mongodb_rest_client::MongoDBRESTClient;

#[derive(Deserialize)]
struct PathParams {
    raw_start: f64
}

/// Registers routes in Actix-web
pub fn configure(cfg: &mut web::ServiceConfig) {
    cfg.service(
        web::scope("/api/telemetry")
            .service(post)
        );
}

#[post("/{raw_start}")]
async fn post(mongodb_rest_client: web::Data<MongoDBRESTClient>, path: web::Path<PathParams>, body: web::Bytes) -> impl Responder {
    let params = path.into_inner();
    let client = mongodb_rest_client.get_client();
    
    let raw_start = params.raw_start;
    let seconds = raw_start as i64;
    let nanos = ((raw_start % 1.0) * 1_000_000_000.0) as u32;
    // Use timestamp_opt instead of timestamp
    let start = Utc.timestamp_opt(seconds, nanos).single().expect("Invalid timestamp");
    let container_time = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs_f64();
    let ssl_handshake = Utc::now().signed_duration_since(start);
    let start_ref = Utc::now();

    let measure = Utc::now();
    let received_data = String::from_utf8_lossy(&body);
    let body: Option<Value> = serde_json::from_str(&received_data).ok().unwrap();
    let deserialization = Utc::now().signed_duration_since(measure);

    let db_name = "test";
    let collection_name = "source";
    let measure = Utc::now();
    let document_id = client.create(db_name, collection_name, body).await.unwrap();
    let create_object = Utc::now().signed_duration_since(measure);

    let measure = Utc::now();
    let document = client.read(db_name, collection_name, document_id).await.unwrap();
    let get_object = Utc::now().signed_duration_since(measure);

    let measure = Utc::now();
    let _json_data = json!(document);
    let serialization = Utc::now().signed_duration_since(measure);

    let measure = Utc::now();
    for _ in 0..15 {
        client.update(db_name, collection_name, document_id, doc! {"time": SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs_f64()}).await.unwrap();
    }
    let high_iops = Utc::now().signed_duration_since(measure);

    let measure = Utc::now();
    client.delete(db_name, collection_name, document_id).await.unwrap();
    let delete_object = Utc::now().signed_duration_since(measure);

    let total = Utc::now().signed_duration_since(start_ref);

    let telemetry = doc! {
        "start": raw_start,
        "container_time": container_time,
        "ssl_handshake": ssl_handshake.num_milliseconds(),
        "deserialization": deserialization.num_milliseconds(),
        "create_object": create_object.num_milliseconds(),
        "get_object": get_object.num_milliseconds(),
        "serialization": serialization.num_milliseconds(),
        "high_iops": high_iops.num_milliseconds(),
        "delete_object": delete_object.num_milliseconds(),
        "total": total.num_milliseconds(),
        "type": "Rust"
    };

    // Return telemetry data as JSON
    HttpResponse::Ok().json(telemetry)
}